internal/provider_framework/attribute_validators.go

```diff
@@ -0,0 +1,219 @@
+	tflog.Debug(ctx, "Validating attribute is set together with other required attributes", map[string]interface{}{
+		"attribute":          attrPathToString(req.AttributePath),
+		"requiredAttributes": av.requiredAttributes,
+	})
+
+	var v attr.Value
+	diags := tfsdk.ValueAs(ctx, req.AttributeConfig, &v)
+	if diags.HasError() {
+		res.Diagnostics.Append(diags...)
+		return
+	}
+
+	for _, path := range av.requiredAttributes {
+		var o attr.Value
+		req.Config.GetAttribute(ctx, path, &o)
+
+		isVNull, diags := isAttrValueNull(ctx, req.AttributePath, v)
+		res.Diagnostics.Append(diags...)
+		if diags.HasError() {
+			return
+		}
+
+		isONull, diags := isAttrValueNull(ctx, path, o)
+		res.Diagnostics.Append(diags...)
+		if diags.HasError() {
+			return
+		}
+
+		if !isVNull && isONull {
+			res.Diagnostics.AddAttributeError(
+				req.AttributePath,
+				fmt.Sprintf("Attribute %q missing", attrPathToString(path)),
+				fmt.Sprintf("%q must be specified when %q is specified", attrPathToString(path), attrPathToString(req.AttributePath)),
+			)
+			return
+		}
+	}
+}
+
+// conflictsWithAttributeValidator checks that a set of *tftypes.AttributePath,
+// including the attribute it's applied to, are not set simultaneously.
+// This implements the validation logic declaratively within the tfsdk.Schema.
+//
+// The provided tftypes.AttributePath must be "absolute",
+// and starting with top level attribute names.
+type conflictsWithAttributeValidator struct {
+	conflictingAttributes []*tftypes.AttributePath
+}
+
+// conflictsWith is a helper to instantiate conflictsWithAttributeValidator.
+func conflictsWith(attributePaths ...*tftypes.AttributePath) tfsdk.AttributeValidator {
+	return &conflictsWithAttributeValidator{attributePaths}
+}
+
+var _ tfsdk.AttributeValidator = (*conflictsWithAttributeValidator)(nil)
+
+func (av conflictsWithAttributeValidator) Description(ctx context.Context) string {
+	return av.MarkdownDescription(ctx)
+}
+
+func (av conflictsWithAttributeValidator) MarkdownDescription(ctx context.Context) string {
+	return fmt.Sprintf("Ensure that if an attribute is set, these are not set: %q", av.conflictingAttributes)
+}
+
+func (av conflictsWithAttributeValidator) Validate(ctx context.Context, req tfsdk.ValidateAttributeRequest, res *tfsdk.ValidateAttributeResponse) {
+	tflog.Debug(ctx, "Validating attribute is not set together with other conflicting attributes", map[string]interface{}{
+		"attribute":             attrPathToString(req.AttributePath),
+		"conflictingAttributes": av.conflictingAttributes,
+	})
+
+	var v attr.Value
+	diags := tfsdk.ValueAs(ctx, req.AttributeConfig, &v)
+	if diags.HasError() {
+		res.Diagnostics.Append(diags...)
+		return
+	}
+
+	for _, path := range av.conflictingAttributes {
+		var o attr.Value
+		req.Config.GetAttribute(ctx, path, &o)
+
+		isVNull, diags := isAttrValueNull(ctx, req.AttributePath, v)
```

Comment:
Looks like the determination of whether `v` is null, is performed on each iteration. Believe this could be moved above the `for` loop as the return value will be stable.